### ISSUE TITLE
Should Residual Be Applied Where masks == True in CogVideoX Transformer?

### ISSUE THREAD

**Issue by Joseph-LIN-Tech** [TencentARC/VideoPainter#30]
I have a question regarding the logic for applying residuals in the `CogVideoXTransformer3D` model.

In the file located at:

VideoPainter/diffusers/src/diffusers/models/transformers/cogvideox_transformer_3d.py

specifically around lines 603 and 609, the code currently uses the following pattern:

    torch.where(masks == False, hidden_states + branch_block_samples[i], hidden_states)

However, based on the mask semantics elsewhere in the codebase — where a mask value of `1` indicates a region that requires inpainting, and `0` indicates a region that should be preserved — it seems that residuals should perhaps be added where `masks == True`, not where `masks == False`.

To clarify:

| Mask Value | Meaning                     | Expected Behavior         |
|------------|------------------------------|----------------------------|
| 1          | Region needs inpainting      | Add residual               |
| 0          | Region already filled/known  | Keep hidden_states unchanged |

Currently, the residual is being applied when `masks == False` (i.e., when the region is already filled), which seems inconsistent with the intended behavior.

Question:
- Is this behavior intentional?
- Or should the condition be updated to `masks == True` so that residuals are only added to regions that require inpainting?

Potential Correction:

    torch.where(masks == True, hidden_states + branch_block_samples[i], hidden_states)

Thanks for your clarification and great work on this project!

**Comment by yxbian23** [comment]
> I have a question regarding the logic for applying residuals in the `CogVideoXTransformer3D` model.
> 
> In the file located at:
> 
> VideoPainter/diffusers/src/diffusers/models/transformers/cogvideox_transformer_3d.py
> 
> specifically around lines 603 and 609, the code currently uses the following pattern:
> 
> ```
> torch.where(masks == False, hidden_states + branch_block_samples[i], hidden_states)
> ```
> 
> However, based on the mask semantics elsewhere in the codebase — where a mask value of `1` indicates a region that requires inpainting, and `0` indicates a region that should be preserved — it seems that residuals should perhaps be added where `masks == True`, not where `masks == False`.
> 
> To clarify:
> 
> Mask Value	Meaning	Expected Behavior
> 1	Region needs inpainting	Add residual
> 0	Region already filled/known	Keep hidden_states unchanged
> Currently, the residual is being applied when `masks == False` (i.e., when the region is already filled), which seems inconsistent with the intended behavior.
> 
> Question:
> 
> * Is this behavior intentional?
> * Or should the condition be updated to `masks == True` so that residuals are only added to regions that require inpainting?
> 
> Potential Correction:
> 
> ```
> torch.where(masks == True, hidden_states + branch_block_samples[i], hidden_states)
> ```
> 
> Thanks for your clarification and great work on this project!

Thank you for taking the time to look into this and for raising the question! 🙏

This behavior is in fact intentional. In CogVideoXTransformer3D our *branch* carries only the **background** information, and our goal is to

1. **preserve** the known background where no inpainting is needed
2. **add back** that preserved background into the backbone (CogVideoX) in exactly those regions

Since in our convention

* `mask == 1` marks the *inpainting* region, and
* `mask == 0` marks the *background* (already known) region,

we must re-add the branch residual only where `mask == False` (i.e. `mask == 0`). That way, the background stays untouched in the masked (inpainting) region, and we “paint” the preserved background back onto the trunk elsewhere.

```python
# preserve/apply residual only in background (mask == False) regions:
hidden_states = torch.where(
    masks == False,
    hidden_states + branch_block_samples[i],
    hidden_states
)
```

For more details on the design rationale and derivation, please see Section 3.2 of our paper, where we explicitly describe how the branch is used to carry and re-inject background information.

Hope this clarifies things—thanks again for your careful reading and for helping us improve the project! If anything remains unclear please let us know.

**Comment by Joseph-LIN-Tech** [comment]
Thanks for your reply!
